flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.store.file.operation;
+
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.table.store.file.Snapshot;
+import org.apache.flink.table.store.file.manifest.ManifestEntry;
+import org.apache.flink.table.store.file.manifest.ManifestFileMeta;
+import org.apache.flink.table.store.file.manifest.ManifestList;
+import org.apache.flink.table.store.file.mergetree.sst.SstPathFactory;
+import org.apache.flink.table.store.file.utils.FileStorePathFactory;
+import org.apache.flink.table.store.file.utils.FileUtils;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Default implementation of {@link FileStoreExpire}. It retains a certain number or period of
+ * latest snapshots.
+ *
+ * <p>NOTE: This implementation will keep at least one snapshot so that users will not accidentally
+ * clear all snapshots.
+ */
+public class FileStoreExpireImpl implements FileStoreExpire {
+
+    // snapshots exceeding any constraint will be expired
+    private final int numRetained;
+    private final long millisRetained;
+
+    private final FileStorePathFactory pathFactory;
+    private final ManifestList manifestList;
+    private final FileStoreScan scan;
+
+    public FileStoreExpireImpl(
+            int numRetained,
+            long millisRetained,
+            FileStorePathFactory pathFactory,
+            ManifestList.Factory manifestListFactory,
+            FileStoreScan scan) {
+        this.numRetained = numRetained;
+        this.millisRetained = millisRetained;
+        this.pathFactory = pathFactory;
+        this.manifestList = manifestListFactory.create();
+        this.scan = scan;
+    }
+
+    @Override
+    public void expire() {
+        Long latestSnapshotId = pathFactory.latestSnapshotId();
+        if (latestSnapshotId == null) {
+            // no snapshot, nothing to expire
+            return;
+        }
+
+        // binary search for the last snapshot to expire
```

Comment:
It is not cost effective to use dichotomous lookups (assuming there are enough surviving snapshots).
Just start traversing from the beginning.